pennylane_lightning/lightning_gpu/lightning_gpu.py

```diff
@@ -0,0 +1,754 @@
+
+            if tape_return_type is State:  # pragma: no cover
+                raise QuantumFunctionError(
+                    "This method does not support statevector return type. "
+                    "Use vjp method instead for this purpose."
+                )
+
+            # Check adjoint diff support
+            self._check_adjdiff_supported_operations(tape.operations)
+
+            processed_data = self._process_jacobian_tape(tape, starting_state, use_device_state)
+
+            if not processed_data:  # training_params is empty
+                return np.array([], dtype=self.state.dtype)
+
+            trainable_params = processed_data["tp_shift"]
+            # pylint: disable=pointless-string-statement
+            """
+            This path enables controlled batching over the requested observables, be they explicit, or part of a Hamiltonian.
+            The traditional path will assume there exists enough free memory to preallocate all arrays and run through each observable iteratively.
+            However, for larger system, this becomes impossible, and we hit memory issues very quickly. the batching support here enables several functionalities:
+            - Pre-allocate memory for all observables on the primary GPU (`batch_obs=False`, default behaviour): This is the simplest path, and works best for few observables, and moderate qubit sizes. All memory is preallocated for each observable, and run through iteratively on a single GPU.
+            - Evenly distribute the observables over all available GPUs (`batch_obs=True`): This will evenly split the data into ceil(num_obs/num_gpus) chunks, and allocate enough space on each GPU up-front before running through them concurrently. This relies on C++ threads to handle the orchestration.
+            - Allocate at most `n` observables per GPU (`batch_obs=n`): Providing an integer value restricts each available GPU to at most `n` copies of the statevector, and hence `n` given observables for a given batch. This will iterate over the data in chnuks of size `n*num_gpus`.
+            """
+            adjoint_jacobian = AdjointJacobianC64() if self.use_csingle else AdjointJacobianC128()
+
+            if self._batch_obs:
+                adjoint_jacobian = adjoint_jacobian.batched
+
+            jac = adjoint_jacobian(
+                processed_data["state_vector"],
+                processed_data["obs_serialized"],
+                processed_data["ops_serialized"],
+                trainable_params,
+            )
+
+            jac = np.array(jac)  # only for parameters differentiable with the adjoint method
+            jac = jac.reshape(-1, len(trainable_params))
+            jac_r = np.zeros((jac.shape[0], processed_data["all_params"]))
+
+            jac_r[:, processed_data["record_tp_rows"]] = jac
+
+            if hasattr(qml, "active_return"):
+                return self._adjoint_jacobian_processing(jac_r) if qml.active_return() else jac_r
```

Comment:
The active return system is now the default (I believe). This can likely be removed, and replaced with the appropriate branch of the above corresponding with up-stream PennyLane.

Reply:
Let's work on it later in a follow-up PR since LKokkos has the similar implementation.